runner/master_log.py

```diff
@@ -0,0 +1,172 @@
+from dataclasses import dataclass, field
+from datetime import datetime
+from typing import Optional, Dict, List
+
+from utils.constants import EventType
+
+
+@dataclass
+class LogEvent:
+    type: EventType
+    timestamp: datetime
+    test_id: Optional[str] = None
+    data: Optional[Dict[str, str]] = field(default_factory=dict)
+
+
+class MasterLog:
+
+    def __init__(self):
+        self.log: List[LogEvent] = []
+
+    def add_send_message(self, message: str, timestamp: datetime, test_id: Optional[str] = None, is_question: Optional[bool] = False):
+        if test_id:
+            event = LogEvent(EventType.SEND_MESSAGE, timestamp, test_id,  {"message": message, "is_question": is_question})
+        else:
+            event = LogEvent(EventType.SEND_FILL, timestamp, data={"message": message, "is_question": False})
+
+        self.add_event(event)
+
+    def add_response_message(self, message: str, timestamp: datetime, test_id: Optional[str] = None, is_question: Optional[bool] = False):
```

Comment:
`Optional` typing means that the value can be `None`. I don't think that's the intended with `is_question`.